ISPN-13421 Add approximate size metrics

danberindei opened the pull request:
https://issues.redhat.com/browse/ISPN-13421

tristantarrant commented:
Needs to be rebased now

pruivo commented on client/hotrod-client/src/main/java/org/infinispan/client/hotrod/ServerStatistics.java:
```diff
@@ -18,17 +18,35 @@
 
    /**
-    * Number of put operations.
+    * Approximate number of entry copies currently in the server cache.
```
I'm not sure what this means :fearful: 
is it the entries on a single server? or cluster-wide? does it include persistency? 

why `Entries owned by multiple nodes are counted once for each owner`? Wouldn't be better to sum everything a divide by the `num-owners`?

ps. I haven't checked impl yet.

pruivo replied:
ps2. why do we need 2 `APPROXIMATE_ENTRIES` and `APPROXIMATE_ENTRIES_UNIQUE`?

danberindei replied:
> is it the entries on a single server? or cluster-wide? 

I kept the old javadoc style, which is modeled on `Stats`.
Because `Stats` is supposed to be used as an interface for both a single server and the cluster, the javadocs don't say explicitly that they apply to either one.

`Encoder2x.statsResponse()` does write cluster-wide statistics with different name (e.g. `approximateEntries` becomes `globalApproximateEntries`), but since they didn't have constants in `ServerStatistics` before I didn't want to add them now. Perhaps I should...

> does it include persistency?

Yes, I'll add that info.

> why Entries owned by multiple nodes are counted once for each owner? Wouldn't be better to sum everything a divide by the num-owners?
> ps2. why do we need 2 APPROXIMATE_ENTRIES and APPROXIMATE_ENTRIES_UNIQUE?

That's the difference between APPROXIMATE_ENTRIES and APPROXIMATE_ENTRIES_UNIQUE: the "unique" stat only counts entries owned as primary owner, so there are no duplicates.

The idea was to let the user know when there's a state transfer in progress and the cache is holding more entry copies than expected.

pruivo commented on client/hotrod-client/src/main/java/org/infinispan/client/hotrod/impl/RemoteCacheImpl.java:
```diff
@@ -282,14 +282,21 @@ public ClientStatistics clientStatistics() {
+      return op.execute()
+            .thenApply(statsMap -> {
+               ServerStatisticsImpl stats = new ServerStatisticsImpl();
+               for (Map.Entry<String, String> entry : statsMap.entrySet()) {
```
offtopic, why not change `StatsOperation` to return `ServerStatistics`? 
`ServerStatistics` creates a `HashMap` and here, we are creating another.

danberindei replied:
It never occurred to me, I just moved the existing code to a lambda :)

danberindei replied:
I changed StatsOperation

pruivo commented on core/src/main/java/org/infinispan/persistence/manager/PassivationPersistenceManager.java:
```diff
@@ -22,6 +25,8 @@
 import io.reactivex.rxjava3.core.Flowable;
 
 public class PassivationPersistenceManager extends DelegatingPersistenceManager {
+   @Inject ComponentRef<InternalDataContainer<?, ?>> dataContainer;
```
unused?

pruivo commented on core/src/main/java/org/infinispan/persistence/manager/PersistenceManagerImpl.java:
```diff
@@ -160,6 +164,19 @@
+      for (StoreStatus storeStatus : stores) {
+         if (storeStatus.characteristics.contains(Characteristic.BULK_READ) &&
+             predicate.test(storeStatus.config)) {
+            firstStoreStatus = storeStatus;
```
remove `firstStoreStatus` var and 
```suggestion
            return storeStatus;
```

pruivo commented on core/src/main/java/org/infinispan/persistence/manager/PersistenceManagerImpl.java:
```diff
@@ -114,6 +117,7 @@
    @Inject ComponentRef<InternalExpirationManager<Object, Object>> expirationManager;
    @Inject DistributionManager distributionManager;
    @Inject InterceptorChainFactory interceptorChainFactory;
+   @Inject ComponentRef<InternalDataContainer<?, ?>> dataContainer;
```
unsed?

pruivo commented on core/src/main/java/org/infinispan/persistence/manager/PersistenceManagerImpl.java:
```diff
@@ -160,6 +164,19 @@
+   private StoreStatus getStoreStatusLocked(Predicate<? super StoreConfiguration> predicate) {
+      StoreStatus firstStoreStatus = null;
+      for (StoreStatus storeStatus : stores) {
+         if (storeStatus.characteristics.contains(Characteristic.BULK_READ) &&
```
a comment saying why do we need `BULK_READ`.

danberindei replied:
It shouldn't have been here, I extracted the method and forgot to compose the predicates in the caller

pruivo commented on core/src/main/java/org/infinispan/persistence/spi/NonBlockingStore.java:
```diff
@@ -580,25 +586,23 @@
     */
    default CompletionStage<Long> approximateSize(IntSet segments) {
-      return size(segments);
+      return CompletableFuture.completedFuture(-1L);
```
```suggestion
      return SIZE_UNAVAILABLE_FUTURE;
```

danberindei replied:
I didn't try to optimize the default implementations thinking stores should override the default, and even if they didn't, one allocation wouldn't matter for a store. I used a constant in PersistenceManagerImpl only because that method can be invoked when there is no store.

wburns replied:
I think it might be useful to add the -1 future to CompletableFutures, but I am fine either way.

pruivo commented on core/src/main/java/org/infinispan/persistence/support/NonBlockingStoreAdapter.java:
```diff
@@ -140,7 +141,7 @@ private String nextTraceId(String operationName) {
    public CompletionStage<Long> approximateSize(IntSet segments) {
       // Old SPI didn't support approximations
-      return size(segments);
+      return CompletableFuture.completedFuture(-1L);
```
```suggestion
      return SIZE_UNAVAILABLE_FUTURE;
```

pruivo commented on core/src/main/java/org/infinispan/interceptors/impl/CacheMgmtInterceptor.java:
```diff
@@ -695,7 +777,7 @@ public int getNumberOfEntriesInMemory() {
+      if (cacheConfiguration.memory().maxSizeBytes() > 0) {
          return dataContainer.evictionSize();
       }
       return 0;
```
Based on `org.infinispan.stats.Stats#getDataMemoryUsed` javadoc, this should return `-1`.

danberindei replied:
Changed

pruivo commented on core/src/main/java/org/infinispan/stats/impl/ClusterCacheStatsImpl.java:
```diff
@@ -88,7 +92,10 @@
    @Inject Configuration cacheConfiguration;
    @Inject GlobalConfiguration globalConfiguration;
-   private ClusterExecutor clusterExecutor;
+   @Inject PersistenceManager persistenceManager;
```
unused

pruivo commented on core/src/main/java/org/infinispan/stats/impl/ClusterCacheStatsImpl.java:
```diff
@@ -261,6 +272,29 @@ public long getMisses() {
+         displayName = "Cluster wide approximate number of cache entries"
+   )
+   public long getApproximateEntries() {
+      return getStatAsInt(APPROXIMATE_ENTRIES);
```
```suggestion
      return getStatAsLong(APPROXIMATE_ENTRIES);
```
same for the methods below.

danberindei replied:
Good catch!

pruivo commented on core/src/main/java/org/infinispan/stats/impl/ClusterCacheStatsImpl.java:
```diff
@@ -494,9 +527,9 @@ private DistributedCacheStatsCallable(String cacheName, boolean accurateSize) {
 
          AdvancedCache<Object, Object> remoteCache =
                SecurityActions.getUnwrappedCache(embeddedCacheManager.getCache(cacheName)).getAdvancedCache();
+         CacheMgmtInterceptor stats = SecurityActions.getCacheMgmtInterceptor(remoteCache);
```
why going through the `ComponentRegistry` instead of `getFirstInterceptorWhichExtends`?

danberindei replied:
Because I didn't notice `getFirstInterceptorWhichExtends` :)

pruivo commented on core/src/main/java/org/infinispan/stats/impl/StatsCollector.java:
```diff
@@ -290,6 +299,30 @@ public long getTimeSinceReset() {
+         displayName = "Approximate number of entries"
+   )
+   public long getApproximateEntries() {
+      return CompletionStages.join(persistenceManager.running().approximateSize(AccessMode.BOTH, SEGMENT_0));
```
why are you using the `persistenceManager`? `SimpleCache` does not support persistence AFAIK.

danberindei replied:
Yep, you're right

pruivo commented on persistence/rocksdb/src/main/java/org/infinispan/persistence/rocksdb/RocksDBStore.java:
```diff
@@ -925,6 +919,18 @@ protected RocksIterator wrapIterator(RocksDB db, ReadOptions readOptions, int se
+         return blockingManager.supplyBlocking(() -> {
+            ColumnFamilyHandle handle = getHandle(0);
+            try {
+               return Long.parseLong(db.getProperty(handle, "rocksdb.estimate-num-keys"));
```
```suggestion
               return Long.parseLong(db.getProperty(defaultColumnFamilyHandle, "rocksdb.estimate-num-keys"));
```

danberindei commented on core/src/main/java/org/infinispan/interceptors/impl/CacheMgmtInterceptor.java:
```diff
@@ -674,10 +684,81 @@ public long getAverageRemoveTimeNanos() {
+                                  if (privateSize >= 0)
+                                     return CompletableFuture.completedFuture(privateSize);
+
+                                  return persistenceManager.running().approximateSize(AccessMode.SHARED, sharedSegments);
```
The idea is that private stores actually store the entries multiple times, so `approximateEntries()` should count all the copies.
Shared stores do not store each entry multiple times, so it would be wrong to count them once for every primary+backup owner.

The other alternative I considered was to only count shared store entries in `approximateEntriesUnique()`, but I thought that might be too surprising with the current stat names. It would have made more sense if I made `approximateEntries` do what `approximateEntriesUnique` does now, with `approximateEntriesLocal` counting the entries stored locally on each node and thus excluding shared stores. But then `approximateEntriesLocal` is jarring when used through `ClusterCacheStats` (which extends `Stats`). Maybe `approximateEntriesPrivate` would work?

pruivo replied:
pls no more stats about entries :/

pruivo replied:
>  Shared stores do not store each entry multiple times, so it would be wrong to count them once for every primary+backup owner.

ok, but the shared store is counting *all* segments for `approximateEntries`. 
if you put 1000 entries, all nodes will report `approximateEntries  = 1000`
It just does not seem correct to me.

danberindei replied:
> pls no more stats about entries :/

I didn't mean to add more entries, I meant renaming
approximateEntries -> approximateEntriesLocal/approximateEntriesPrivate
approximateEntriesUnique -> approximateEntries

danberindei replied:
> > Shared stores do not store each entry multiple times, so it would be wrong to count them once for every primary+backup owner.
> 
> ok, but the shared store is counting _all_ segments for `approximateEntries`. if you put 1000 entries, all nodes will report `approximateEntries = 1000` It just does not seem correct to me.

No, `sharedSegments` is `distributionManager.getCacheTopology().getLocalWriteSegments()` for dist caches.

pruivo replied:
> No, sharedSegments is distributionManager.getCacheTopology().getLocalWriteSegments() for dist caches.

ah! my bad. When I read `approximatePersistenceSize(allSegments, writeSegments)`, I "assumed" `allSegments` were used by the shared store :)

pruivo replied:
`approximateEntries` -> `approximateLocalEntries`
`approximateEntriesUnique` -> `approximateLocalEntriesUnique`
may I suggest ^ ? 
having the word `local` in the stat name makes everything clearer IMO.

danberindei replied:
@pruivo I stayed away from "local" in the name because `Stats` and `ClusterCacheStats` must use the same method names to refer to local stats (in StatsImpl) and to cluster-wide stats (in ClusterCacheStatsImpl), and although `CacheMgmtInterceptor` doesn't implement the interface, the rule of thumb seems to be to use the same names.

I could keep the `approximateEntries` and `approximateEntriesUnique` names in `Stats` + rename the methods to `approximateLocalEntries` and `approximateLocalEntriesUnique` in `CacheMgmtInterceptor`, but I'm afraid the difference (which doesn't exist for other stats) might make the names more confusing instead of making them clearer.

OTOH, as you've noticed, Hot Rod's `ServerStatistics` already uses different names for the cluster-wide stats, and the REST API doesn't seem to provide cluster-wide stats at the moment, so maybe this is a good time to walk away from using the `Stats` interface and the same stat names for both local and cluster-wide stats.

pruivo replied:
fair enough. I'll wait for @wburns green light before merging.

danberindei commented on core/src/main/java/org/infinispan/interceptors/impl/CacheMgmtInterceptor.java:
```diff
@@ -674,10 +684,81 @@ public long getAverageRemoveTimeNanos() {
+   }
+
+   private long approximateTotalSize(long persistenceSize, IntSet segments) {
+      long inMemorySize = dataContainer.sizeIncludingExpired(segments);
```
Actually `OffHeapDataContainer` and `SegmentedBoundedOffHeapDataContainer` are O(1) (since the number of segments is constant).

For local caches with eviction, `BoundedOffHeapDataContainer` does iterate over the entries, so I'll implement it in O(1).

wburns replied:
We could always optimize the BoundedOffHeapDataContainer by checking if the segments provided = the total segment count in the config and just return the size of the map.

danberindei replied:
> We could always optimize the BoundedOffHeapDataContainer by checking if the segments provided = the total segment count in the config and just return the size of the map.

@wburns I tried to do this, but then I wanted to do the same check for all BoundedOffHeapDataContainer methods that receive a `segments` parameter and I noticed that the assertion was failing in some tests.

When there is a segmented store, computing the size involves calling publishEntries(segment) once for each segment, even though the cache is local, so I created https://issues.redhat.com/browse/ISPN-13574 to remove that special case and always use a single segment.

danberindei commented:
Updated @pruivo, please have another look at it.

pruivo reviewed (approved):
LGTM. Hopefully, didn't miss anything important :)

wburns commented on core/src/main/java/org/infinispan/persistence/manager/PersistenceManagerImpl.java:
```diff
@@ -836,64 +840,86 @@ private boolean containedInAdapter(NonBlockingStore nonBlockingStore, String ada
-         checkStoreAvailability();
-         if (log.isTraceEnabled()) {
-            log.tracef("Obtaining size from stores");
+         if (!isEnabled() || !isAvailable()) {
```
This should work outside the read lock as the variables are volatile.

danberindei replied:
I think for availability I would have to check again after acquiring the read lock.
But I could move the `isEnabled()` check out, for the benefit of caches that will never have a store.

wburns commented on core/src/main/java/org/infinispan/persistence/manager/PersistenceManagerImpl.java:
```diff
@@ -836,64 +840,86 @@ private boolean containedInAdapter(NonBlockingStore nonBlockingStore, String ada
+
+         // Ignore stores without BULK_READ, they don't implement approximateSize()
+         StoreStatus firstStoreStatus = getStoreStatusLocked(storeStatus -> {
+            return storeStatus.hasCharacteristic(Characteristic.BULK_READ) &&
```
Don't need brackets.

danberindei replied:
Without brackets IntelliJ formats the code "the wrong way" (with my settings, at least, because I have `ij_java_align_multiline_parameters_in_calls = true` in my EditorConfig

wburns commented on core/src/main/java/org/infinispan/persistence/manager/PersistenceManagerImpl.java:
```diff
@@ -836,64 +840,86 @@ private boolean containedInAdapter(NonBlockingStore nonBlockingStore, String ada
+                  .thenApply(size -> {
+                     // Counting only the keys in the given segments would be expensive,
+                     // so we compute an estimate assuming that each segment has a similar number of entries
+                     LocalizedCacheTopology cacheTopology = distributionManager.getCacheTopology();
```
Nice :)

wburns commented on core/src/main/java/org/infinispan/persistence/manager/PersistenceManagerStub.java:
```diff
@@ -34,6 +34,8 @@
 @Scope(Scopes.NAMED_CACHE)
 @SurvivesRestarts
 public class PersistenceManagerStub implements PersistenceManager {
+   private static final CompletableFuture<Long> SIZE_ZERO_FUTURE = CompletableFuture.completedFuture(0L);
```
Seems probably useful to promote this to CompletableFutures, wdyt?

danberindei replied:
This is the only user so far: for some old-forgotten reason, `PersistenceManagerImpl` returns -1L when there's no store, and `PersistenceManagerStub` returns 0L all the time.

Actually Gustavo removed all usages of `PersistenceManagerStub` from production code, so I think I'll just delete the class.

wburns commented on core/src/main/java/org/infinispan/persistence/sifs/Index.java:
```diff
@@ -209,12 +209,11 @@ public void deleteFileAsync(int fileId) {
-         size += seg.size.get();
-         if (size < 0) {
-            return Long.MAX_VALUE;
+      for (int s = 0; s < segments.length; s++) {
```
Unfortunately, this segment is not the same as a cache segment. The SIFS uses its own notion of segment to reduce B-tree overhead by splitting it into its own "segments" keyed by the hashCode.

We probably need to just do the same calculation that PersistenceMangerImpl does when segmented.

danberindei replied:
Thanks, I will also add a test in `BaseNonBlockingStoreTest`

wburns reviewed (commented):
Overall looks, great. However, I am worried about all of the possible overflow issues we have in all the stats part of the code. I guess we had those issues before, though and it isn't needed for this PR.

danberindei commented:
Updated

wburns commented on core/src/main/java/org/infinispan/persistence/sifs/NonBlockingSoftIndexFileStore.java:
```diff
@@ -510,7 +516,10 @@ protected boolean isIndexLoaded() {
+      long totalSize = index.approximateSize();
+      IntSet matchingSegments = IntSets.mutableCopyFrom(segments);
+      matchingSegments.retainAll(matchingSegments);
+      return CompletableFuture.completedFuture(totalSize * matchingSegments.size() / ownedSegments.size());
```
I would say we can use the TemporaryTable to store how many segments it owns instead, but I am fine with this impl.

danberindei replied:
Done

wburns reviewed (approved):
LGTM pending CI. Can you squash some of the commits as well @danberindei ?

danberindei commented:
Squashed

wburns commented on core/src/main/java/org/infinispan/persistence/sifs/NonBlockingSoftIndexFileStore.java:
```diff
@@ -510,7 +510,11 @@ protected boolean isIndexLoaded() {
-      return CompletableFuture.completedFuture(index.approximateSize());
+      long totalSize = index.approximateSize();
+      IntSet matchingSegments = IntSets.mutableCopyFrom(segments);
+      matchingSegments.retainAll(matchingSegments);
```
You mean `matchingSegment.retainAll(temporaryTable.getOwnedSegments()) ?

danberindei replied:
Oops!

danberindei replied:
Fixed

wburns reviewed (approved):
LGTM, just waiting on CI.

pruivo commented:
`testApproximateSize` is failing for JDBC testsuite.

danberindei commented:
The JDBC test suite was failing because of ISPN-13636: BaseJdbcStore.size(segments) ignores the segments parameter
I added a workaround in JdbcStringBasedStore

wburns commented:
> The JDBC test suite was failing because of ISPN-13636: BaseJdbcStore.size(segments) ignores the segments parameter I added a workaround in JdbcStringBasedStore

The reason it ignores it is because the base one doesn't support segmentation. The JdbcStringBasedStore needs to override size and implement it itself instead.

pruivo commented:
@danberindei there is a rocksdb failure that might be related:
```
Caused by: org.rocksdb.RocksDBException: lock hold by current process, acquire time 1642554224 acquiring thread 140459746715392: /home/infinispan/workspace/Infinispan_PR-9693/persistence/rocksdb/target/infinispanTempFiles/RocksDBStoreFunctionalTest/data/testRemoveCacheWithPassivation/data/LOCK: No locks available
```

danberindei commented:
Sorry @pruivo, I introduced that failure when I removed `PersistenceManagerSub` and I made `TrackingPersistenceManager` extend `DelegatingPersistenceManager`. When replacing the component, the original `PersistenceManagerImpl.start()` would be called again, and fail because of the RocksDB lock.

To make up for my mistake, in addition to this test I also fixed for a random failure in `JdbcStringBasedCacheStorePassivation.testSoftRestartWithPassivation`, which assumed passivation is synchronous ;)

pruivo reviewed (approved):
LGTM.

danberindei commented:
There was a conflict with ISPN-13680, it's fixed now. 

The failed build was because I was rebasing all my WIP branches and accidentally included a commit from one of them.

wburns commented on persistence/jdbc/src/main/java/org/infinispan/persistence/jdbc/stringbased/JdbcStringBasedStore.java:
```diff
@@ -247,13 +258,26 @@ protected void extraStopSteps() {
+   @Override
+   public CompletionStage<Long> approximateSize(IntSet segments) {
+      // The size method ignores segments, see ISPN-13636
+      return size(segments).thenApply(totalSize -> {
```
```suggestion
      return size(segments);
```
I don't see any reason to do the retainAll here. The segments should be handled in the size method directly.

danberindei replied:
See the comment on line 263 :D

wburns replied:
I saw it, but there is no reason to do all this extra code IMO. The size method should be the one pushing this completely. The size method is already broken if you want to get approximateSize more accurate I would just say put this code in size is my point.

danberindei replied:
I see what you mean now. 

At the time I thought you were working on a fix for `size()` and I didn't want to mess with it, but nowadays I'm thinking maybe the proper fix for ISPN-13636 is to remove the `segments` parameter from `NonBlockingStore.size()`.

The method is only really used in one place, `PersistenceManagerImpl.size(Predicate)`, which always calls it with the full set of segments, so it would be wasteful to write e.g. a SQL query that counts the rows for a set of segments that will only be used in the tests.

wburns replied:
Well the intent of `size(IntSet)` on a store is to eventually optimize the Cache.size() when it has a store without passivation so it doesn't need to even ask in memory and just asks the store, but that hasn't been done yet.

danberindei replied:
Fair enough, I moved the approximation to `size()`

wburns commented on query/src/main/java/org/infinispan/query/clustered/QueryPartitioner.java:
```diff
@@ -39,7 +39,7 @@ public QueryPartitioner(AdvancedCache<?, ?> cache) {
       segmentsPerMember.put(localAddress, bitSet);
       for (int s = 0; s < numSegments; s++) {
-         if (!localSegments.contains(s)) {
+         if (!bitSet.get(s)) {
```
We should probably fix line 39 as well. Should be `localSegments.forEach((IntConsumer) bitSet::set);` to avoid possible boxing and unboxing.

wburns reviewed (commented):
Just a couple minor things.

danberindei commented:
Updated, CI looks good.

wburns commented:
Integrated into main, thanks @danberindei !

wburns commented on core/src/main/java/org/infinispan/persistence/manager/PersistenceManagerImpl.java:
```diff
@@ -836,64 +837,89 @@ private boolean containedInAdapter(NonBlockingStore nonBlockingStore, String ada
+         }
+         if (firstStoreStatus.hasCharacteristic(Characteristic.SEGMENTABLE)) {
+            return firstStoreStatus.store.approximateSize(segments)
+                                         .whenComplete((ignore, ignoreT) -> releaseReadLock(stamp));
```
Actually just realized this line is not in the else, so the read lock isn't released :(

wburns replied:
I am fixing it in my PR https://github.com/infinispan/infinispan/pull/9845

danberindei replied:
Oops, you're quite right. I wonder why we have no test failures, though...

wburns replied:
Assuming we just don't have a test that hits that line for w/e reason.
